### ISSUE TITLE
Cannot compare Long values within body

### ISSUE THREAD

**Issue by fpompermaier** [rest-assured/rest-assured#741]
As already faced in https://groups.google.com/forum/#!topic/rest-assured/mf-VIHM-aJA, a body element cannot be compared to a long value.

I think the best solution could be to allow function passing (as for headers) in order to convert the value to a long. E.g:

```java
long expected = 1L;
when().get("/test/{id}", expected).
then().body("id",Long::valueOf, Matchers.is(expected));
```

At the moment I have to compare the id value with an integer version of  **expected** that is definitely unsafe...

**Comment by johanhaleby** [comment]
Yes I've been thinking about function passing as well. The problem is that the api already supports the use of [path arguments](https://github.com/rest-assured/rest-assured/wiki/Usage#path-arguments) which will make the API twice as big if adding functions as well. I have been thinking about making "arguments" more general so that it would be possible to supply a "mapper function argument". Maybe like this:

```java
then().body("id", map(Long::valueOf), Matchers.is(expected));
```

Where `map` returns another implementation of the `Argument` class (probably `Argument` would implement another interface that can be common for all types (`map`, `withArgs` etc)).

**Comment by fpompermaier** [comment]
Fine for me, the important thing is to fix this somehow . However, from my point of view, adding another sign accepting functions wouldn't be a big problem (but it's just my personal opinion).

**Comment by johanhaleby** [comment]
What do you mean by sign? You mean an overloaded method? It will become a huge burden for me and it will increase the API so much that people will have a hard time figuring out which methods to call I think.

**Comment by fpompermaier** [comment]
Yes I mean exactly that. At the moment I solved the problem transforming the entire body to the response class and checking with assertions the object state. E.g.:

```java
MyClass c = .... 
        then().
           extract().response().as(MyClass.class);
Assert.equals(c.getId(), expected)
```

**Comment by johanhaleby** [comment]
Yeah, I really would have liked something like this as well. Now all I need is to find some free time working on this :)

**Comment by mcantrell** [comment]
Ran into the same issue and solved it with a custom matcher. In my case, Groovy equality works out well so I made a GroovyEquals matcher. Maybe you could include a similar matcher with the framework:



    protected ExtractableResponse uploadFile(String workspaceId, File file) {
        return given()
                .multiPart("file", file)
        //when
                .when()
                .post("/workspaces/${workspaceId}/files")
        //then
                .then()
                .statusCode(equalTo(200))
                .body("workspaceId", is(workspaceId))
                .body("name", is(file.name))
                .body("size", new GroovyEquals(file.length()))
        // response
                .extract()

    }


    class GroovyEquals extends BaseMatcher {
        protected final Object value

        GroovyEquals(Object value) {
            this.value = value
        }

        @Override
        boolean matches(Object item) {
            return item == value
        }

        @Override
        void describeTo(Description description) {
            description.appendText("groovyEquals(${value})")
        }
    }

**Comment by kyxap** [comment]
To avoid this issue you can use `equalTo` instead of `is`:

```java
import static org.hamcrest.Matchers.*;

...
.body(".dueDate", equalTo(invoice.getDueDate))
...
```

**Comment by think01** [comment]
I still have this problem, maybe even worse: if I use an integer value for the comparison, the test keeps failing because it casts the json value to 0.0. If I put 0.0 or 0.0f as expected value, it fail with "expected 0.0f, found 0". Apparently it always uses the wrong cast.

I resolved as suggested above:

````
        MyResponseClass res =
        given().... when()....then()
        .extract().response().as(MyResponseClass.class);
        Assert.assertEquals(res.getValue(), new BigDecimal(0));
````

**Comment by johanhaleby** [comment]
Hmm that's strange. There are multiple test cases in RA that checks for this and the build would not pass if they were to fail. What version of RA are you using and what version of Groovy?

**Comment by think01** [comment]
Rest Assured 3.0.7
About Groovy, not sure: I'm building a Java project with Gradle build and I have no direct dependencies on it. I see groovy-2.4.15.jar in my classpath.

**Comment by johanhaleby** [comment]
Could you share the code of `MyResponseClass`?

**Comment by johanhaleby** [comment]
If "value" in MyResponseClass is of type Object then I suppose this could happen. You should change it to "float" if this is what you expect.

**Comment by think01** [comment]
It's of type BigDecimal.

**Comment by think01** [comment]
The above code is working.

My original test (with the erroneous behavior described earlier) is:

```
given()....then()
.body("resultCode", equalTo(0)); // <-- resultCode is integer in JSON/Swagger specs
.body("availableQty", equalTo(0f)); // <--availableQty is number in JSON/Swagger specs
```

**Comment by johanhaleby** [comment]
@think01 Well then it necessarily has to be BigDecimal when you compare it. This actually has nothing to do with rest assured. `assertEquals` is from JUnit and once you've mapped to a Java object you're outside the realm of REST Assured.

**Comment by johanhaleby** [comment]
Could you log the response and show it here? Just add `log().all()` after `then()`.

**Comment by think01** [comment]
Sorry, I was writing the response here but I suddenly had to leave.

Here it is the testing code snippet:
```
Response r = given()....when().post("/qtaslot");
ValidatableResponse vr = r.then().log().all()
.statusCode(200)
.body("resultCode", equalTo(0))
.body("disponibilitaCorrente", equalTo(0f));
```
And the log is:
```
HTTP/1.1 200 
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Date: Mon, 23 Apr 2018 09:35:20 GMT

{
    "resultCode": 0,
    "errorMessage": null,
    "disponibilitaCorrente": 0,
    "carico": 0,
    "scarico": 0,
    "unitaMisura": null
}
```
Test fails with:
```
java.lang.AssertionError: 1 expectation failed.
JSON path disponibilitaCorrente doesn't match.
Expected: <0.0F>
  Actual: 0
```

**Comment by think01** [comment]
But also (another test to same api rest):
```
Response r = given()....when().post("/qtaslot");
ValidatableResponse vr = r.then().log().all()
.statusCode(200)
.body("resultCode", equalTo(0))
.body("carico", equalTo(99.0f))
.body("scarico", equalTo(99.0f))
.body("disponibilitaCorrente", equalTo(0));
```

Log is:
```HTTP/1.1 200 
Content-Type: application/json;charset=UTF-8
Transfer-Encoding: chunked
Date: Mon, 23 Apr 2018 12:27:28 GMT

{
    "resultCode": 0,
    "errorMessage": null,
    "disponibilitaCorrente": 0.0000,
    "carico": 99.0000,
    "scarico": 99.0000,
    "unitaMisura": "PZ"
}
```

Test fails with:
```
java.lang.AssertionError: 1 expectation failed.
JSON path disponibilitaCorrente doesn't match.
Expected: <0>
  Actual: 0.0
```

**Comment by think01** [comment]
Ok, maybe I got it: some tests return 0 (requiring to test against an integer), while others 0.0000 (requiring to test against a float). I changed all values together with same type so I went back and forth from one failing test to another, not realizing they were different tests.

Thx for the support.

**Comment by AnatoliiStepaniuk** [comment]
So. still getting this error in 2019. 
```
JSON path [0].score doesn't match.
Expected: 42
  Actual: 42
```

**Comment by cynthux** [comment]
Same issue here:

```
java.lang.AssertionError: 1 expectation failed.
JSON path client_id doesn't match.
Expected: 123
  Actual: 123
```

As @AnatoliiStepaniuk said, the value in json is treated as int although the model had a Long type. My workaround was to parse the result as follows:

`.body("id", equalTo(Integer.valueOf(id.toString())));`

where id is a Long value.

**Comment by rajeevpmr** [comment]
I did like this and it worked
 body("product.subTotal",equalTo(115.0f)); Not sure if this is the right way, a good workaround though.

**Comment by jwedel** [comment]
I'm also seeing this issue:

```
java.lang.AssertionError: 1 expectation failed.
JSON path foo doesn't match.
Expected: <42L>
  Actual: 42
```

Any updates here? :)

**Comment by aschoerk** [comment]
Perhaps the same problem as: https://stackoverflow.com/questions/57397579/how-to-do-hamcrest-number-comparisons-without-regarding-the-class

**Comment by matheusmessora** [comment]
Same problem here:
Expected: is "1"
     but: was <1>

**Comment by davide-imbriaco** [comment]
same problem here

```
Expected: <69443L>
  Actual: 69443
```

my code is

```
.body("data.id", equalTo(record_id)) // record_id is type long
```

data model is `long`, but rest assured map this to an int when reading response

a workaround could be having a `equalToNumber()` that convert number types as appropriate when comparing, es:

```
.body("data.id", equalToNumber(record_id)) // record_id is type long
```

or a family of `equalToLong|Float|...()` that enforce type conversion before comparing

**Comment by michelrisucci** [comment]
This is a Hamcrest matching issue and could be easily fixed using Hamcrest `ComparatorMatcherBuilder`.

Assuming this JSON as REST response, **note the "salary" property having integer and double values**:
```json
[
  {
    "id": 1,
    "name": "Richard Miles",
    "age": 30,
    "salary": 1234.56
  },
  {
    "id": 2,
    "name": "Mary Stiles",
    "age": 25,
    "salary": 2500
  },
  {
    "id": 3,
    "name": "John Noakes",
    "age": 20
  }
]
```

To safely compare them, try to create a custom matcher that parses the values BEFORE comparing:
```java
import org.hamcrest.Matcher;
import org.hamcrest.comparator.ComparatorMatcherBuilder;

// ComparatorMatcherBuilder is a special Hamcrest class that builds custom comparing Matchers
final Matcher<Number> greaterThanOrEqualToZeroDouble = ComparatorMatcherBuilder
		// Custom comparing rule: double parse the values to compare
		.comparedBy(Comparator.comparingDouble(Number::doubleValue))
		// Business rule (customizable): validate if matching values are greater than or equals to ZERO
		.greaterThanOrEqualTo(0D);

given("https://base.url.to")
		.get("/people")
		.then()
		.body("salary", everyItem(anyOf(greaterThanOrEqualToZeroDouble, nullValue())));
```

As it is not a RestAssured issue, I think this could end this issue.

**Comment by machadoit** [comment]
Also run into some problems trying to compare with Longs. I was trying to compare an Array<Long> relying on .hasItems, and the problem is due to Long... For now I am forcing Integer, and it works, but it is a bit "wrong" since the API that I am testing does indeed return Longs.

**Comment by ohtaeg** [comment]
```
java.lang.AssertionError: 1 expectation failed.
JSON path id doesn't match.
Expected: <1L>
  Actual: <1>
```

any update here~? TT

**Comment by CamilYed** [comment]
any update here? TT

**Comment by gian1200** [comment]
Looks similar to this https://github.com/hamcrest/JavaHamcrest/issues/220

**Comment by welovelain** [comment]
Expected: <11048002L>
  Actual: <11048002>
  
  Any updates here? :))

**Comment by javar0ne** [comment]
Had the same problem and wrote a simple Hamcrest matcher.
I'm using pattern matching, which is present from Java 14 on, but you can easily replace it with manual cast.
It converts the `actualValue` parameter of `matches` method to Long or Long array and then calls the parent `matches` method.

```java
package com.javar0ne.app.test;

import org.hamcrest.Matcher;
import org.hamcrest.core.IsEqual;

import java.lang.reflect.Array;

public class EqualToLong<T> extends IsEqual<T> {
    public EqualToLong(T equalArg) {
        super(equalArg);
    }

    @Override
    public boolean matches(Object actualValue) {
        Object actualValueParsed = getActualValueParsed(actualValue);

        return super.matches(actualValueParsed);
    }

    private Object getActualValueParsed(Object actualValue) {
        Object actualValueParsed;

        if(actualValue == null) return null;

        if(actualValue.getClass().isArray()) {
            int actualValueLength = Array.getLength(actualValue);
            actualValueParsed = new Object[actualValueLength];

            for(int i = 0; i < actualValueLength; i++) {
                Object currentValue = Array.get(actualValue, i);

                Array.set(actualValueParsed, i, parseLong(currentValue));
            }
        } else {
            actualValueParsed = parseLong(actualValue);
        }

        return actualValueParsed;
    }

    private Long parseLong(Object actualValue) {
        if(actualValue instanceof Integer intValue) {
            return intValue.longValue();
        } else if(actualValue instanceof String strValue) {
            return Long.parseLong(strValue);
        }

        throw new IllegalArgumentException("Wrong type for actualValue, cannot parse to Long");
    }

    public static <T> Matcher<T> equalToLong(T operand) {
        return new EqualToLong<>(operand);
    }
}
```
